File: bindings/go/backend/workflow.go
package backend

import (
	"context"
	"time"
)

type WorkflowStepDefinition struct {
	Name                  string
	ExpectedExecutionTime time.Duration
	MaxRetries            uint32
}

type WorkflowDefinition struct {
	WorkflowName          string
	WorkflowVersion       int64
	StartStep             string
	ExpectedExecutionTime time.Duration
	Singleton             bool
	Steps                 []WorkflowStepDefinition
	Metadata              JSON
}

type RegisterWorkflowDefinitionRequest struct {
	Definition WorkflowDefinition
}

type WorkflowDefinitionRegistration struct {
	Definition WorkflowDefinition
	Created    bool
}

type WorkflowExecutionRecord struct {
	WorkflowExecutionID string
	WorkflowName        string
	WorkflowVersion     int64
	CurrentStep         string
	Status              string
	Attempt             uint64
	ClaimedBy           *string
	ClaimExpiresAt      *time.Time
	State               JSON
}

type StartWorkflowRequest struct {
	WorkflowExecutionID string
	WorkflowName        string
	WorkflowVersion     int64
	StartStep           string
	State               JSON
}

type ClaimWorkflowStepRequest struct {
	WorkflowExecutionID string
	WorkflowName        string
	WorkflowVersion     int64
	Worker              string
	Now                 time.Time
	LeaseDuration       time.Duration
	MaxSteps            uint32
}

type KeepAliveWorkflowStepRequest struct {
	WorkflowExecutionID string
	Worker              string
	CurrentStep         string
	Now                 time.Time
	LeaseDuration       time.Duration
}

type CompleteWorkflowStepRequest struct {
	WorkflowExecutionID string
	Worker              string
	CompletedStep       string
	NextStep            *string
	State               JSON
}

type FailWorkflowStepRequest struct {
	WorkflowExecutionID string
	Worker              string
	FailedStep          string
	Reason              string
	Now                 time.Time
	MaxAttempts         uint32
}

type CancelWorkflowRequest struct {
	WorkflowExecutionID string
	Reason              string
}

type WorkflowStore interface {
	RegisterDefinition(ctx context.Context, backend Backend, request RegisterWorkflowDefinitionRequest) (WorkflowDefinitionRegistration, error)

	RegisterDefinitionTx(ctx context.Context, tx Transaction, request RegisterWorkflowDefinitionRequest) (WorkflowDefinitionRegistration, error)

	InspectDefinition(ctx context.Context, backend Backend, workflowName string, workflowVersion int64) (*WorkflowDefinition, error)

	InspectDefinitionTx(ctx context.Context, tx Transaction, workflowName string, workflowVersion int64) (*WorkflowDefinition, error)

	Start(ctx context.Context, backend Backend, request StartWorkflowRequest) (WorkflowExecutionRecord, error)

	StartTx(ctx context.Context, tx Transaction, request StartWorkflowRequest) (WorkflowExecutionRecord, error)

	ClaimSteps(ctx context.Context, backend Backend, request ClaimWorkflowStepRequest) ([]WorkflowExecutionRecord, error)

	ClaimStepsTx(ctx context.Context, tx Transaction, request ClaimWorkflowStepRequest) ([]WorkflowExecutionRecord, error)

	KeepAliveStep(ctx context.Context, backend Backend, request KeepAliveWorkflowStepRequest) (WorkflowExecutionRecord, error)

	KeepAliveStepTx(ctx context.Context, tx Transaction, request KeepAliveWorkflowStepRequest) (WorkflowExecutionRecord, error)

	CompleteStep(ctx context.Context, backend Backend, request CompleteWorkflowStepRequest) (WorkflowExecutionRecord, error)

	CompleteStepTx(ctx context.Context, tx Transaction, request CompleteWorkflowStepRequest) (WorkflowExecutionRecord, error)

	FailStep(ctx context.Context, backend Backend, request FailWorkflowStepRequest) (WorkflowExecutionRecord, error)

	FailStepTx(ctx context.Context, tx Transaction, request FailWorkflowStepRequest) (WorkflowExecutionRecord, error)

	Cancel(ctx context.Context, backend Backend, request CancelWorkflowRequest) (WorkflowExecutionRecord, error)

	CancelTx(ctx context.Context, tx Transaction, request CancelWorkflowRequest) (WorkflowExecutionRecord, error)

	Inspect(ctx context.Context, backend Backend, workflowExecutionID string) (*WorkflowExecutionRecord, error)

	InspectTx(ctx context.Context, tx Transaction, workflowExecutionID string) (*WorkflowExecutionRecord, error)
}

File: bindings/rust/workflow.rs
use std::time::{Duration, SystemTime};

use crate::backend::{Backend, BackendResult};
use crate::json::Json;

#[derive(Debug, Clone)]
pub struct WorkflowStepDefinition {
    pub name: String,
    pub expected_execution_time: Duration,
    pub max_retries: u32,
}

#[derive(Debug, Clone)]
pub struct WorkflowDefinition {
    pub workflow_name: String,
    pub workflow_version: i64,
    pub start_step: String,
    pub expected_execution_time: Duration,
    pub singleton: bool,
    pub steps: Vec<WorkflowStepDefinition>,
    pub metadata: Json,
}

#[derive(Debug, Clone)]
pub struct RegisterWorkflowDefinitionRequest {
    pub definition: WorkflowDefinition,
}

#[derive(Debug, Clone)]
pub struct WorkflowDefinitionRegistration {
    pub definition: WorkflowDefinition,
    pub created: bool,
}

#[derive(Debug, Clone)]
pub struct WorkflowExecutionRecord {
    pub workflow_execution_id: String,
    pub workflow_name: String,
    pub workflow_version: i64,
    pub current_step: String,
    pub status: String,
    pub attempt: u64,
    pub claimed_by: Option<String>,
    pub claim_expires_at: Option<SystemTime>,
    pub state: Json,
}

#[derive(Debug, Clone)]
pub struct StartWorkflowRequest {
    pub workflow_execution_id: String,
    pub workflow_name: String,
    pub workflow_version: i64,
    pub start_step: String,
    pub state: Json,
}

#[derive(Debug, Clone)]
pub struct ClaimWorkflowStepRequest {
    pub workflow_execution_id: String,
    pub workflow_name: String,
    pub workflow_version: i64,
    pub worker: String,
    pub now: SystemTime,
    pub lease_duration: Duration,
    pub max_steps: u32,
}

#[derive(Debug, Clone)]
pub struct KeepAliveWorkflowStepRequest {
    pub workflow_execution_id: String,
    pub worker: String,
    pub current_step: String,
    pub now: SystemTime,
    pub lease_duration: Duration,
}

#[derive(Debug, Clone)]
pub struct CompleteWorkflowStepRequest {
    pub workflow_execution_id: String,
    pub worker: String,
    pub completed_step: String,
    pub next_step: Option<String>,
    pub state: Json,
}

#[derive(Debug, Clone)]
pub struct FailWorkflowStepRequest {
    pub workflow_execution_id: String,
    pub worker: String,
    pub failed_step: String,
    pub reason: String,
    pub now: SystemTime,
    pub max_attempts: u32,
}

#[derive(Debug, Clone)]
pub struct CancelWorkflowRequest {
    pub workflow_execution_id: String,
    pub reason: String,
}

pub trait WorkflowStore<B: Backend> {
    fn register_definition(
        &self,
        backend: &B,
        request: &RegisterWorkflowDefinitionRequest,
    ) -> BackendResult<WorkflowDefinitionRegistration>;

    fn register_definition_tx(
        &self,
        tx: &mut B::Tx,
        request: &RegisterWorkflowDefinitionRequest,
    ) -> BackendResult<WorkflowDefinitionRegistration>;

    fn inspect_definition(
        &self,
        backend: &B,
        workflow_name: &str,
        workflow_version: i64,
    ) -> BackendResult<Option<WorkflowDefinition>>;

    fn inspect_definition_tx(
        &self,
        tx: &mut B::Tx,
        workflow_name: &str,
        workflow_version: i64,
    ) -> BackendResult<Option<WorkflowDefinition>>;

    fn start(
        &self,
        backend: &B,
        request: &StartWorkflowRequest,
    ) -> BackendResult<WorkflowExecutionRecord>;

    fn start_tx(
        &self,
        tx: &mut B::Tx,
        request: &StartWorkflowRequest,
    ) -> BackendResult<WorkflowExecutionRecord>;

    fn claim_steps(
        &self,
        backend: &B,
        request: &ClaimWorkflowStepRequest,
    ) -> BackendResult<Vec<WorkflowExecutionRecord>>;

    fn claim_steps_tx(
        &self,
        tx: &mut B::Tx,
        request: &ClaimWorkflowStepRequest,
    ) -> BackendResult<Vec<WorkflowExecutionRecord>>;

    fn keep_alive_step(
        &self,
        backend: &B,
        request: &KeepAliveWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord>;

    fn keep_alive_step_tx(
        &self,
        tx: &mut B::Tx,
        request: &KeepAliveWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord>;

    fn complete_step(
        &self,
        backend: &B,
        request: &CompleteWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord>;

    fn complete_step_tx(
        &self,
        tx: &mut B::Tx,
        request: &CompleteWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord>;

    fn fail_step(
        &self,
        backend: &B,
        request: &FailWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord>;

    fn fail_step_tx(
        &self,
        tx: &mut B::Tx,
        request: &FailWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord>;

    fn cancel(
        &self,
        backend: &B,
        request: &CancelWorkflowRequest,
    ) -> BackendResult<WorkflowExecutionRecord>;

    fn cancel_tx(
        &self,
        tx: &mut B::Tx,
        request: &CancelWorkflowRequest,
    ) -> BackendResult<WorkflowExecutionRecord>;

    fn inspect(
        &self,
        backend: &B,
        workflow_execution_id: &str,
    ) -> BackendResult<Option<WorkflowExecutionRecord>>;

    fn inspect_tx(
        &self,
        tx: &mut B::Tx,
        workflow_execution_id: &str,
    ) -> BackendResult<Option<WorkflowExecutionRecord>>;
}
